Dev documentation (#284)
getupio-undistro/undistro: website/src/util/docs/index.md

&nbsp;

&nbsp;

# 11 - Development
## How to setup the development environment

The first step in getting involved with Undistro is to download the source code, and since the development is carried out by means of the Git version control system, it can be used to clone the [repository](https://github.com/getupio-undistro/undistro). If you don't already have Git installed, it can be found in most package managers as well as in the official downloads [page](https://git-scm.com/downloads). Furthermore, a Github account will also be required for sending changes back, so be sure to have one.

> Both backend and frontend software lies on the same repository. The second is mostly composed by the _UI_ folder, while the first by the _api_, _charts_, _cmd_ and _tilt\_modules_ folders and also by many files in the root directory. Such folders will be discussed with more details later on.

Reviewer: ```suggestion
Both backend and frontend software lies on the same repository. The second is mostly composed by the _UI_ folder, while the first is by the _apis_, _controllers_, _pkg_,_cmd_ and _tilt\_modules_, and so on. Such folders will be discussed with more details later on.
```

Reviewer: Here may be interesting to talk about the charts folder in a separate way. What u think?

 

### Backend

Undistro's backed is mostly written in the Go programming language. Go is a concise, statically typed, compiled language with a powerful concurrency model, also being the language in which Kubernetes is written. Most software repositories have a package for Go, but any case, it can also be obtained in [this page](https://golang.org/dl/).

> Another crucial piece of software to write code for Undistro is Helm, which is a component manager for Kubernetes, providing an unified way to share, track and upgrade cluster's setups. Helm is distributed as a single statically linked binary, which can be obtained [here](https://helm.sh/docs/intro/install/).

Reviewer: ```suggestion
Another crucial piece of software to write code for Undistro is Helm, which is a package manager for Kubernetes, providing a unified way to share, track and upgrade cluster setups. Helm is distributed as a single statically linked binary, which can be obtained [here](https://helm.sh/docs/intro/install/).
```

Next, a local Kubernetes cluster must be created. Any Kubernetes distribution should suffice, but KinD and Minikube have been the ones where most testing has occurred. KinD stands for "Kubernetes in Docker", being a tool used to run Kubernetes clusters locally through Docker containers. Head over to its [website](https://kind.sigs.k8s.io/) for installation instructions. Minikube is an advanced tool used to run local Kubernetes clusters which supports multiple runtimes such as Virtual Machines, Container engines or the host OS. It also supports LoadBalancers and plugins for a diverse set of purposes. Check out Minikube's [documentation](https://minikube.sigs.k8s.io/docs/start/) for installation instructions.

KinD has the advantage of speed, being very quick to create and execute clusters. On the other hand, Minikube wins at extensibility, its multiple runtime options permit a finer-grained setup, and the plugins greatly simplify its configuration. Choose one of them to proceed with the setup.

> In a similar way, Docker is another one of the requirements. Docker is a container runtime used to package programs and their dependencies in an image, making it very portable and isolating its execution from the host operating system. Install it with your package manager or through Docker's get started [page](https://www.docker.com/get-started).

Reviewer: ```suggestion
In a similar way, Docker is another one of the requirements to set up UnDistro locally. Docker is a container runtime used to package programs and their dependencies in an image, making it very portable and isolating its execution from the host operating system. Install it with your package manager or through Docker's get started [page](https://www.docker.com/get-started).
```
